### PR TITLE
Interactions Command Localization

### PR DESCRIPTION
**Pull request by Cenngo**
This PR adds Application Command localization to Core, Rest and WebSocket projects. Introduces a `ILocalizationManger` interface and default Localization Managers for Json and ResX in preperation for the Slash Commands localization feature release.

Two LocalizationManagers ship with this update, JsonLocalizationManager and ResxLocalizationManager.

JsonLocaliationManager uses a nested data structure similar to Discord's Application Commands schema. You can get the Json schema [here](https://gist.github.com/Cenngo/d46a881de24823302f66c3c7e2f7b254). LocalizationManager accepts a base path and a base file name and automatically discovers every resource file ( `\basePath\fileName.locale.json`  ). 

```json
{
    "command_1":{
        "name": "localized_name",
        "description": "localized_description",
        "parameter_1":{
            "name": "localized_name",
            "description": "localized_description"
        }
    },
    "group_1":{
        "name": "localized_name",
        "description": "localized_description",
        "command_1":{
            "name": "localized_name",
             "description": "localized_description",
             "parameter_1":{
                 "name": "localized_name",
                  "description": "localized_description"
            },
            "parameter_2":{
                 "name": "localized_name",
                  "description": "localized_description"
            }
        }
    }
}
```

~~ResxLocalizationManager uses the [naming scheme](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization?view=aspnetcore-6.0#resource-file-naming) used by ASP.NET ( `group1\command1\parameter2.fr-FR.resx` or  `group1.command1.parameter2.fr-FR.resx` ). Every resx file must have two fields: name and description. Individual .resx files can be defined for every group/command/parameter. Space characters within Context command names are tokenized with tilde character.~~

The initial ResxLocalizationManager concept is abandoned due to how wasteful the pattern would get in larger scale bots.

ResxLocalizationManager uses `.` delimited key names to traverse the resource files and get the localized strings (`group1.group2.command.parameter.name`). Every key path must end with either `.name` or `.description` including parameter choice strings. [Discord.Tools.LocalizationTemplate.Resx](https://www.nuget.org/packages/Discord.Tools.LocalizationTemplate.Resx) dotnet tool can be used to create localization file templates.

Interaction Service can be easily configured to use any `ILocalizationManager` implementation by setting the LocalizationManager field of InteractionServiceConfig.

## TODO
- [x] Add per-request header support to rest client(#2261)
- [x] Add localization support to parameter choices
- [x] Add locale parameter to Get*ApplicationCommandsAsync methods for fetching localized name and description strings

## Breaking Changes
```diff
- A new requestHeaders parameter is added to every Send* method of IRestClient.
- Two new parameters have been added to every `Get*ApplicationCommandsAsync` method (`withLocalizations`, `locale`)
```

### CODE REVIEW

File: src/Discord.Net.Core/Entities/Guilds/IGuild.cs [hunk 1]
```diff
@@ -1194,12 +1194,17 @@ public interface IGuild : IDeletable, ISnowflakeEntity
         /// <summary>
         ///     Gets this guilds application commands.
         /// </summary>
+        /// <param name="withLocalizations">
+        ///     Whether to include full localization dictionaries in the returned objects,
+        ///     instead of the localized name and description fields.
+        /// </param>
+        /// <param name="locale">The target locale of the localized name and description fields. Sets <c>X-Discord-Locale</c> header, which takes precedence over <c>Accept-Language</c>.</param>
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Guilds/IGuild.cs]:
```suggestion
        /// <param name="locale">The target locale of the localized name and description fields. Sets the <c>X-Discord-Locale</c> header, which takes precedence over <c>Accept-Language</c>.</param>
```

File: src/Discord.Net.Core/Entities/Interactions/ContextMenus/MessageCommandBuilder.cs [hunk 2]
```diff
@@ -86,6 +97,30 @@ public MessageCommandBuilder WithDefaultPermission(bool isDefaultPermission)
             return this;
         }
 
+        /// <summary>
+        ///     Sets the <see cref="NameLocalizations"/> collection.
+        /// </summary>
+        /// <param name="nameLocalizations">Localization dictionary for the name field of this command.</param>
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/ContextMenus/MessageCommandBuilder.cs]:
```suggestion
        /// <param name="nameLocalizations">The localization dictionary to use for the name field of this command.</param>
```

File: src/Discord.Net.Core/Entities/Interactions/ContextMenus/UserCommandBuilder.cs [hunk 3]
```diff
@@ -84,6 +95,30 @@ public UserCommandBuilder WithDefaultPermission(bool isDefaultPermission)
             return this;
         }
 
+        /// <summary>
+        ///     Sets the <see cref="NameLocalizations"/> collection.
+        /// </summary>
+        /// <param name="nameLocalizations">Localization dictionary for the name field of this command.</param>
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/ContextMenus/UserCommandBuilder.cs]:
```suggestion
        /// <param name="nameLocalizations">The localization dictionary to use for the name field of this command.</param>
```

File: src/Discord.Net.Core/Entities/Interactions/IApplicationCommandOption.cs [hunk 4]
```diff
@@ -61,5 +61,31 @@ public interface IApplicationCommandOption
         ///     Gets the allowed channel types for this option.
         /// </summary>
         IReadOnlyCollection<ChannelType> ChannelTypes { get; }
+
+        /// <summary>
+        ///     Gets the localization dictionary for the name field of this command.
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/IApplicationCommandOption.cs]:
```suggestion
        ///     Gets the localization dictionary for the name field of this command option.
```

File: src/Discord.Net.Core/Entities/Interactions/IApplicationCommandOption.cs [hunk 5]
```diff
@@ -61,5 +61,31 @@ public interface IApplicationCommandOption
...
+        /// </summary>
+        IReadOnlyDictionary<string, string> NameLocalizations { get; }
+
+        /// <summary>
+        ///     Gets the localization dictionary for the description field of this command.
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/IApplicationCommandOption.cs]:
```suggestion
        ///     Gets the localization dictionary for the description field of this command option.
```

File: src/Discord.Net.Core/Entities/Interactions/IApplicationCommandOption.cs [hunk 6]
```diff
@@ -61,5 +61,31 @@ public interface IApplicationCommandOption
...
+        /// </summary>
+        IReadOnlyDictionary<string, string> DescriptionLocalizations { get; }
+
+        /// <summary>
+        ///     Gets the localized name of this command.
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/IApplicationCommandOption.cs]:
```suggestion
        ///     Gets the localized name of this command option.
```

File: src/Discord.Net.Core/Entities/Interactions/IApplicationCommandOption.cs [hunk 7]
```diff
@@ -61,5 +61,31 @@ public interface IApplicationCommandOption
...
+        /// </summary>
+        /// <remarks>
+        ///     Only returned when the `withLocalizations` query parameter is set to <see langword="false"/> when requesting the command.
+        /// </remarks>
+        string NameLocalized { get; }
+
+        /// <summary>
+        ///     Gets the localized description of this command.
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/IApplicationCommandOption.cs]:
```suggestion
        ///     Gets the localized description of this command option.
```

File: src/Discord.Net.Core/Entities/Interactions/IApplicationCommandOptionChoice.cs [hunk 8]
```diff
@@ -14,5 +16,18 @@ public interface IApplicationCommandOptionChoice
         ///     Gets the value of the choice.
         /// </summary>
         object Value { get; }
+
+        /// <summary>
+        ///     Gets the localization dictionary for the name field of this command.
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/IApplicationCommandOptionChoice.cs]:
```suggestion
        ///     Gets the localization dictionary for the name field of this command option.
```

File: src/Discord.Net.Core/Entities/Interactions/IApplicationCommandOptionChoice.cs [hunk 9]
```diff
@@ -14,5 +16,18 @@ public interface IApplicationCommandOptionChoice
...
+        /// </summary>
+        IReadOnlyDictionary<string, string> NameLocalizations { get; }
+
+        /// <summary>
+        ///     Gets the localized name of this command.
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/IApplicationCommandOptionChoice.cs]:
```suggestion
        ///     Gets the localized name of this command option.
```

File: src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs [hunk 10]
```diff
@@ -266,6 +278,116 @@ public SlashCommandBuilder AddOptions(params SlashCommandOptionBuilder[] options
             Options.AddRange(options);
             return this;
         }
+
+        /// <summary>
+        ///     Sets the <see cref="NameLocalizations"/> collection.
+        /// </summary>
+        /// <param name="nameLocalizations">Localization dictionary for the name field of this command.</param>
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs]:
```suggestion
        /// <param name="nameLocalizations">The localization dictionary to use for the name field of this command.</param>
```

File: src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs [hunk 11]
```diff
@@ -266,6 +278,116 @@ public SlashCommandBuilder AddOptions(params SlashCommandOptionBuilder[] options
...
+            }
+
+            _nameLocalizations = new Dictionary<string, string>(nameLocalizations);
+            return this;
+        }
+
+        /// <summary>
+        ///     Sets the <see cref="DescriptionLocalizations"/> collection.
+        /// </summary>
+        /// <param name="descriptionLocalizations">Localization dictionary for the name field of this command.</param>
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs]:
```suggestion
        /// <param name="descriptionLocalizations">The localization dictionary to use for the description field of this command.</param>
```

File: src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs [hunk 12]
```diff
@@ -499,57 +640,62 @@ public SlashCommandOptionBuilder AddOptions(params SlashCommandOptionBuilder[] o
...
-            return AddChoiceInternal(name, value);
+            return AddChoiceInternal(name, value, nameLocalizations);
         }
 
         /// <summary>
         ///     Adds a choice to the current option.
         /// </summary>
         /// <param name="name">The name of the choice.</param>
         /// <param name="value">The value of the choice.</param>
+        /// <param name="nameLocalizations">Localization dictionary for the description field of this command.</param>
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs]:
```suggestion
        /// <param name="nameLocalizations">The localization dictionary to use for the name field of this command option choice.</param>
```

File: src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs [hunk 13]
```diff
@@ -499,57 +640,62 @@ public SlashCommandOptionBuilder AddOptions(params SlashCommandOptionBuilder[] o
...
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs]:
```suggestion
        /// <param name="nameLocalizations">The localization dictionary to use for the name field of this command option choice.</param>
```

File: src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs [hunk 14]
```diff
@@ -679,5 +826,107 @@ public SlashCommandOptionBuilder WithType(ApplicationCommandOptionType type)
             Type = type;
             return this;
         }
+
+        /// <summary>
+        ///     Sets the <see cref="NameLocalizations"/> collection.
+        /// </summary>
+        /// <param name="nameLocalizations">Localization dictionary for the name field of this command.</param>
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs]:
```suggestion
        /// <param name="nameLocalizations">The localization dictionary to use for the name field of this command option.</param>
```

File: src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs [hunk 15]
```diff
@@ -679,5 +826,107 @@ public SlashCommandOptionBuilder WithType(ApplicationCommandOptionType type)
...
+            }
+
+            _nameLocalizations = new Dictionary<string, string>(nameLocalizations);
+            return this;
+        }
+
+        /// <summary>
+        ///     Sets the <see cref="DescriptionLocalizations"/> collection.
+        /// </summary>
+        /// <param name="descriptionLocalizations">Localization dictionary for the description field of this command.</param>
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs]:
```suggestion
        /// <param name="descriptionLocalizations">The localization dictionary to use for the description field of this command option.</param>
```

File: src/Discord.Net.Interactions/LocalizationManagers/ILocalizationManager.cs [hunk 16]
```diff
@@ -0,0 +1,32 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Discord.Interactions
+{
+    /// <summary>
+    ///     Respresents localization provider for Discord Application Commands.
```
Comment (csmir) [inline src/Discord.Net.Interactions/LocalizationManagers/ILocalizationManager.cs]:
```suggestion
    ///     Respresents a localization provider for Discord Application Commands.
```

File: src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs [hunk 17]
```diff
@@ -499,57 +640,62 @@ public SlashCommandOptionBuilder AddOptions(params SlashCommandOptionBuilder[] o
         /// </summary>
         /// <param name="name">The name of the choice.</param>
         /// <param name="value">The value of the choice.</param>
+        /// <param name="nameLocalizations">Localization dictionary for the description field of this command.</param>
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs]:
```suggestion
        /// <param name="nameLocalizations">The localization dictionary for to use the name field of this command option choice.</param>
```

File: src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs [hunk 18]
```diff
@@ -499,57 +640,62 @@ public SlashCommandOptionBuilder AddOptions(params SlashCommandOptionBuilder[] o
...
-            return AddChoiceInternal(name, value);
+            return AddChoiceInternal(name, value, nameLocalizations);
         }
 
         /// <summary>
         ///     Adds a choice to the current option.
         /// </summary>
         /// <param name="name">The name of the choice.</param>
         /// <param name="value">The value of the choice.</param>
+        /// <param name="nameLocalizations">Localization dictionary for the description field of this command.</param>
```
Comment (csmir) [inline src/Discord.Net.Core/Entities/Interactions/SlashCommands/SlashCommandBuilder.cs]:
```suggestion
        /// <param name="nameLocalizations">The localization dictionary for to use the name field of this command option choice.</param>
```

**Review by csmir — CHANGES_REQUESTED** [review]

**Comment by MrCakeSlayer** [comment]
@Cenngo perhaps include some docs/guide for this update as well?

**Review by csmir — COMMENTED** [review]
XML lgtm, but awaiting to approve until I've verified functionality. Going try to build an app on localization for testing

**Comment by Cenngo** [comment]
> XML lgtm, but awaiting to approve until I've verified functionality. Going try to build an app on localization for testing

i also need to do that myself. its been a while since i lasted tested this PR and im sure discord broke a thing or two since then

**Review by csmir — APPROVED** [review]

**Pull request merged**